comm/src/main/scala/coop/rchain/comm/network.scala

```diff
@@ -0,0 +1,115 @@
+package coop.rchain.comm
+
+import java.net.SocketTimeoutException
+import scala.util.{Failure, Success, Try}
+import scala.collection.concurrent
+import scala.concurrent.{Await, Promise}
+import scala.concurrent.duration.{Duration, MILLISECONDS}
+import coop.rchain.kademlia.PeerTable
+
+/**
+  * Implements the lower levels of the network protocol.
+  */
+case class UnicastNetwork(id: NodeIdentifier, endpoint: Endpoint) extends ProtocolHandler {
+
+  case class PendingKey(remote: Seq[Byte], timestamp: Long, seq: Long)
+
+  val pending =
+    new concurrent.TrieMap[PendingKey, Promise[Try[ProtocolMessage]]]
+
+  val local = new ProtocolNode(id, endpoint, this)
+  val comm = UnicastComm(local)
+  val table = PeerTable(local)
+
+  private val receiver = new Thread {
+    override def run =
+      while (true) {
+        comm.recv match {
+          case Success(res) =>
+            for {
+              msg <- ProtocolMessage.parse(res)
+            } dispatch(msg)
+          case Failure(ex: SocketTimeoutException) => ()
+          case Failure(ex)                         => ex.printStackTrace
+        }
+      }
+  }.start
+
+  private def dispatch(msg: ProtocolMessage) =
+    for {
+      sender <- ProtocolMessage.sender(msg)
+    } {
+      table.observe(new ProtocolNode(sender, this))
+      msg match {
+        case ping@PingMessage(_, _) => handlePing(sender, ping)
+        case pong@PongMessage(_, _) => handlePong(sender, pong)
+        case _ => ???
+      }
+    }
+
+  /**
+    * Validate incoming PING and send responding PONG.
+    */
+  private def handlePing(sender: PeerNode, ping: PingMessage) =
+    for {
+      pong <- ProtocolMessage.pong(local, ping)
+    } {
+      comm.send(ProtocolMessage.toBytes(pong), sender)
+    }
+
+  /**
+    * Validate incoming PONG and complete its pending promise.
+    */
+  private def handlePong(sender: PeerNode, pong: PongMessage) =
+    for {
+      ret <- ProtocolMessage.returnHeader(pong)
+      promise <- pending.get(PendingKey(sender.key, ret.timestamp, ret.seq))
+    } {
+      try {
+        promise.success(Success(pong))
+      } catch {
+        case ex: java.lang.IllegalStateException => () // Future already completed
+      }
+    }
+
+  /**
+    * Broadcast a message to all peers in the Kademlia table.
+    */
+  override def broadcast(msg: ProtocolMessage): Seq[Try[Unit]] = {
```

Comment:
Not yet, since I don't use `broadcast` yet, but I suspect that it's the same reason that I'm not spawning threads on behalf of the caller at this level.

Reply:
However, I'll keep this in mind when I see how I use it later.